_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+```
+
+We now run the same code as above, starting from `(20,20)` this time.
+
+```{r}
+num_iterations <- 3
+
+x_star <- torch_tensor(c(20, 0), requires_grad = TRUE)
+
+optimizer <- optim_lbfgs(x_star)
+
+calc_loss <- function() {
+
+  optimizer$zero_grad()
+
+  value <- flower(x_star)
+  cat("Value is: ", as.numeric(value), "\n")
+
+  value$backward()
+  cat("Gradient is: ", as.matrix(x_star$grad), "\n")
+  
+  cat("X is: ", as.matrix(x_star), "\n\n")
+  
+  value
+
+}
+
+for (i in 1:num_iterations) {
+  cat("Iteration: ", i, "\n")
+  optimizer$step(calc_loss)
+}
+```
+
+    Iteration:  1 
+    Value is:  28.28427 
+    Gradient is:  0.8071069 0.6071068 
+    X is:  20 20 
+
+    ...
+    ...
+
+    Value is:  19.33546 
+    Gradient is:  0.8100872 0.6188223 
+    X is:  12.957 14.68274 
+
+    ...
+    ...
+
+    Value is:  18.29546 
+    Gradient is:  0.8096464 0.622064 
+    X is:  12.14691 14.06392 
+
+    ...
+    ...
+
+    Value is:  9.853705 
+    Gradient is:  0.7546976 0.7025688 
+    X is:  5.763702 8.895616 
+
+    Value is:  2635.866 
+    Gradient is:  -0.7407354 -0.6717985 
+    X is:  -1949.697 -1773.551 
+
+    Iteration:  2 
+    Value is:  1333.113 
+    Gradient is:  -0.7413024 -0.6711776 
+    X is:  -985.4553 -897.5367 
+
+    Value is:  30.16862 
+    Gradient is:  -0.7903821 -0.6266789 
+    X is:  -21.02814 -21.72296 
+
+    Value is:  1281.39 
+    Gradient is:  0.7544561 0.6563575 
+    X is:  964.0121 843.7817 
+
+    Value is:  628.1306 
+    Gradient is:  0.7616636 0.6480014 
+    X is:  475.7051 409.7372 
+
+    Value is:  4965690 
+    Gradient is:  -0.7493951 -0.662123 
+    X is:  -3721262 -3287901 
+
+    Value is:  2482306 
+    Gradient is:  -0.7503822 -0.6610042 
+    X is:  -1862675 -1640817 
+
+    Value is:  8.61863e+11 
+    Gradient is:  0.7486113 0.6630091 
+    X is:  645200412672 571423064064 
+
+    Value is:  430929412096 
+    Gradient is:  0.7487153 0.6628917 
+    X is:  322643460096 285659529216 
+
+    Value is:  Inf 
+    Gradient is:  0 0 
+    X is:  -2.826342e+19 -2.503904e+19 
+
+    Iteration:  3 
+    Value is:  Inf 
+    Gradient is:  0 0 
+    X is:  -2.826342e+19 -2.503904e+19 
+
+This has been less of a success. At first, loss decreases nicely, but suddenly, the estimate dramatically overshoots, and keeps bouncing between negative and positive outer space ever after.
+
+Luckily, there is something we can do.
+
+### L-BFGS with line search
+
+Taken in isolation, what a Quasi-Newton method like L-BFGS does is determine the best descent direction. However, as we just saw, a good direction is not enough. With the flower function, wherever we are, the optimal path leads to disaster if we stay on it long enough. We thus need an algorithm that carefully evaluates not only where to go, but also, how far.
+
+For this reason, L-BFGS implementations commonly incorporate line search, that is, a set of rules indicating whether a proposed step length is a good one, or should be improved upon.
```

Comment:
```suggestion
For this reason, L-BFGS implementations commonly incorporate _line search_, that is, a set of rules indicating whether a proposed step length is a good one, or should be improved upon.
```

Reply:
I like giving a term some sort of visual emphasis when you're defining it in a sentence (especially when that's not the only thing going on in the sentence).